### ISSUE TITLE
Data corruption since generic_file_splice_read -> filemap_splice_read change (6.5 compat, but occurs on 6.4 too)

### ISSUE THREAD

**Issue by ckane** [openzfs/zfs#15163]
### System information
Type | Version/Name
 --- | ---
Distribution Name	| Arch
Distribution Version	| Rolling release
Kernel Version	| 6.4.8, 6.5rc1/2/3/4
Architecture	| x86-64
OpenZFS Version	| [commit 36261c8](https://github.com/openzfs/zfs/commit/36261c8238df462b214854ccea1df4f060cf0995)

### Describe the problem you're observing
After the recent changes to get OpenZFS compiling/running on 6.5, there appears to be a possible lingering data corruption bug. In the repeatable example below, it reliably inserts a long run of NULL bytes into a file, causing a build to fail (conveniently, the build of ZFS).

My expectation is that the bug probably exists for any kernel where `filemap_splice_read` exists, which recently has replaced `generic_file_splice_read` in other Linux filesystem code.

### Describe how to reproduce the problem
**Again - despite demonstrating the problem with the OpenZFS build, the problem only manifests itself when running on the ZFS branch at the commit listed above. It just so happens that I'm able to use our build to reproduce the bug.**

1. You need to be running ZFS patched up to the commit listed above. I have reproduced this on Kernel 6.4.8 and all 6.5 RC's up to rc4
2. `git clone https://github.com/openzfs/zfs.git`
3. `cd ./zfs`
4. `./autogen.sh`
5. `mkdir -p ../zfs-test`
6. `cd ../zfs-test`
7. `../zfs/configure --with-linux=/usr/src/linux` (or wherever your headers/source tree is)

Eventually, the `configure` will fail with the following message:
```
configure: error: 
                *** This kernel does not include the required loadable module
                *** support!
                ***
                *** To build OpenZFS as a loadable Linux kernel module
                *** enable loadable module support by setting
                *** `CONFIG_MODULES=y` in the kernel configuration and run
                *** `make modules_prepare` in the Linux source tree.
                ***
                *** If you don't intend to enable loadable kernel module
                *** support, please compile OpenZFS as a Linux kernel built-in.
                ***
                *** Prepare the Linux source tree by running `make prepare`,
                *** use the OpenZFS `--enable-linux-builtin` configure option,
                *** copy the OpenZFS sources into the Linux source tree using
                *** `./copy-builtin <linux source directory>`,
                *** set `CONFIG_ZFS=y` in the kernel configuration and compile
                *** kernel as usual.
```

I enter the directory of the failing test:
```
cd build/config_modules
```

Looking at the `config_modules.c` file, which is resulting in the failure:
```c
/* confdefs.h */
#define PACKAGE_NAME "zfs"
#define PACKAGE_TARNAME "zfs"
#define PACKAGE_VERSION "2.2.99"
#define PACKAGE_STRING "zfs 2.2.99"
#define PACKAGE_BUGREPORT ""
#define PACKAGE_URL ""
#define ZFS_META_NAME "zfs"
#define ZFS_META_VERSION "2.2.99"
#define SPL_META_VERSION ZFS_META_VERSION
#define ZFS_META_RELEASE "1"
#define SPL_META_RELEASE ZFS_META_RELEASE
#define ZFS_META_LICENSE "CDDL"
#define ZFS_META_ALIAS "zfs-2.2.99-1"
#define SPL_META_ALIAS ZFS_META_ALIAS
#define ZFS_META_AUTHOR "OpenZFS"
#define ZFS_META_KVER_MIN "3.10"
#define ZFS_META_KVER_MAX "6.4"
#define PACKAGE "zfs"
#define VERSION "2.2.99"
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
#include <linux/module.h>

                #if !defined(CONFIG_MODULES)
                #error CONFIG_MODULES not defined
                #endif


int
main (void)
{

        ;
        return 0;
}
```

Curiously, this bug does not manifest frequently and the system largely appears to run stable in many other use cases.

The section of `configure` that writes that file looks like below. Including this snippet here as it might help illuminate what conditions need to be true to trigger the bug:
```bash
cat confdefs.h - <<_ACEOF >build/config_modules/config_modules.c

#include <linux/module.h>

                #if !defined(CONFIG_MODULES)
                #error CONFIG_MODULES not defined
                #endif


int
main (void)
{

        ;
        return 0;
}

MODULE_DESCRIPTION("conftest");
MODULE_AUTHOR(ZFS_META_AUTHOR);
MODULE_VERSION(ZFS_META_VERSION "-" ZFS_META_RELEASE);
MODULE_LICENSE("Dual BSD/GPL");

_ACEOF
```

### Include any warning/errors/backtraces from the system logs
There are no errors reported to the console or in kernel messages

**Comment by rincebrain** [comment]
The commit message for torvalds/linux@101df45e7ec36f470559c8fdab8e272cb991ef42 seems like it might have a clue.

     nfsd_splice_actor() has a clause in its loop that chops up a compound page
     into individual pages such that if the same page is seen twice in a row, it
     is discarded the second time.  This is a problem with the advent of
     shmem_splice_read() as that inserts zero_pages into the pipe in lieu of
     pages that aren't present in the pagecache.

**Comment by ckane** [comment]
Thanks for the tips - I'll give `copy_splice_read` a whirl and see if that works more reliably for us

**Comment by ckane** [comment]
> The commit message for [torvalds/linux@101df45](https://github.com/torvalds/linux/commit/101df45e7ec36f470559c8fdab8e272cb991ef42) seems like it might have a clue.
> 
> ```
>  nfsd_splice_actor() has a clause in its loop that chops up a compound page
>  into individual pages such that if the same page is seen twice in a row, it
>  is discarded the second time.  This is a problem with the advent of
>  shmem_splice_read() as that inserts zero_pages into the pipe in lieu of
>  pages that aren't present in the pagecache.
> ```
> 
> Might not, but I came upon that when looking for filemap_splice_read related fixes and went "that sounds familiar"
> 
> Also, at least gfs2 just for now went with `copy_splice_read` instead, which obviously doesn't save us the memory copy, but.
> 
> Either way, I think the semantics of this are different - I think `copy_splice_read` more aptly mimics the old expectations if we don't want to go reworking what we expect to do here.

Thanks again, making this change to `copy_splice_read` appears to at least satisfy the example failure case I was encountering.

**Comment by ckane** [comment]
Fixed in #15164
